// @ng2-dynamic-forms/core/src/service/dynamic-form.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var dynamic_form_array_model_1 = require("../model/form-array/dynamic-form-array.model");
var dynamic_form_group_model_1 = require("../model/form-group/dynamic-form-group.model");
var dynamic_checkbox_group_model_1 = require("../model/checkbox/dynamic-checkbox-group.model");
var dynamic_checkbox_model_1 = require("../model/checkbox/dynamic-checkbox.model");
var dynamic_input_model_1 = require("../model/input/dynamic-input.model");
var dynamic_radio_group_model_1 = require("../model/radio/dynamic-radio-group.model");
var dynamic_select_model_1 = require("../model/select/dynamic-select.model");
var dynamic_slider_model_1 = require("../model/slider/dynamic-slider.model");
var dynamic_switch_model_1 = require("../model/switch/dynamic-switch.model");
var dynamic_textarea_model_1 = require("../model/textarea/dynamic-textarea.model");
var utils_1 = require("../utils");
var DynamicFormService = (function () {
    function DynamicFormService(formBuilder, NG_VALIDATORS, NG_ASYNC_VALIDATORS) {
        this.formBuilder = formBuilder;
        this.NG_VALIDATORS = NG_VALIDATORS;
        this.NG_ASYNC_VALIDATORS = NG_ASYNC_VALIDATORS;
    }
    DynamicFormService.prototype.getCustomValidatorFn = function (validatorName) {
        var validatorFn;
        if (this.NG_VALIDATORS) {
            validatorFn = this.NG_VALIDATORS.find(function (validator) { return validatorName === validator.name; });
        }
        if (!utils_1.isDefined(validatorFn) && this.NG_ASYNC_VALIDATORS) {
            validatorFn = this.NG_ASYNC_VALIDATORS.find(function (asyncValidator) { return validatorName === asyncValidator.name; });
        }
        return validatorFn;
    };
    DynamicFormService.prototype.getValidatorFn = function (validatorName, validatorArgs) {
        var validatorFn = forms_1.Validators[validatorName] || this.getCustomValidatorFn(validatorName);
        if (!utils_1.isFunction(validatorFn)) {
            throw new Error("validator \"" + validatorName + "\" is not provided via NG_VALIDATORS or NG_ASYNC_VALIDATORS");
        }
        return utils_1.isDefined(validatorArgs) ? validatorFn(validatorArgs) : validatorFn;
    };
    DynamicFormService.prototype.getValidators = function (config) {
        var _this = this;
        return utils_1.isDefined(config) ?
            Object.keys(config).map(function (validatorName) { return _this.getValidatorFn(validatorName, config[validatorName]); }) : [];
    };
    DynamicFormService.prototype.createFormArray = function (model) {
        var formArray = [];
        for (var i = 0; i < model.size; i += 1) {
            formArray.push(this.createFormGroup(model.get(i).group));
        }
        return this.formBuilder.array(formArray, this.getValidators(model.validator)[0] || null, this.getValidators(model.asyncValidator)[0] || null);
    };
    DynamicFormService.prototype.createFormGroup = function (group, groupExtra) {
        var _this = this;
        if (groupExtra === void 0) { groupExtra = null; }
        var formGroup = {};
        group.forEach(function (model) {
            if (model.type === dynamic_form_array_model_1.DYNAMIC_FORM_CONTROL_TYPE_ARRAY) {
                var arrayModel = model;
                formGroup[model.id] = _this.createFormArray(arrayModel);
            }
            else if (model.type === dynamic_form_group_model_1.DYNAMIC_FORM_CONTROL_TYPE_GROUP || model.type === dynamic_checkbox_group_model_1.DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP) {
                var groupModel = model, groupExtra_1 = {
                    validator: _this.getValidators(groupModel.validator)[0] || null,
                    asyncValidator: _this.getValidators(groupModel.asyncValidator)[0] || null
                };
                formGroup[model.id] = _this.createFormGroup(groupModel.group, groupExtra_1);
            }
            else {
                var valueControlModel = model;
                formGroup[valueControlModel.id] = new forms_1.FormControl({
                    value: valueControlModel.value,
                    disabled: valueControlModel.disabled
                }, forms_1.Validators.compose(_this.getValidators(valueControlModel.validators || [])), forms_1.Validators.composeAsync(_this.getValidators(valueControlModel.asyncValidators || [])));
            }
        });
        return this.formBuilder.group(formGroup, groupExtra);
    };
    DynamicFormService.prototype.createFormArrayGroup = function (formArrayModel) {
        return this.createFormGroup(formArrayModel.addGroup().group);
    };
    DynamicFormService.prototype.addFormArrayGroup = function (formArray, model) {
        formArray.push(this.createFormArrayGroup(model));
    };
    DynamicFormService.prototype.insertFormArrayGroup = function (index, formArray, model) {
        formArray.insert(index, this.createFormGroup(model.insertGroup(index).group));
    };
    DynamicFormService.prototype.removeFormArrayGroup = function (index, formArray, model) {
        formArray.removeAt(index);
        model.removeGroup(index);
    };
    DynamicFormService.prototype.clearFormArray = function (formArray, model) {
        while (formArray.length > 0) {
            this.removeFormArrayGroup(0, formArray, model);
        }
    };
    DynamicFormService.prototype.findById = function (id, group) {
        return group.find(function (controlModel) { return controlModel.id === id; });
    };
    DynamicFormService.prototype.fromJSON = function (json) {
        var _this = this;
        var formModel = [];
        json.forEach(function (object) {
            switch (object["type"]) {
                case dynamic_form_array_model_1.DYNAMIC_FORM_CONTROL_TYPE_ARRAY:
                    object["groups"].forEach(function (groupObject) { return groupObject["group"] = _this.fromJSON(groupObject["group"]); });
                    object["createGroup"] = function () { return _this.fromJSON(object["originGroup"]); };
                    formModel.push(new dynamic_form_array_model_1.DynamicFormArrayModel(object, object["cls"]));
                    break;
                case dynamic_checkbox_model_1.DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX:
                    formModel.push(new dynamic_checkbox_model_1.DynamicCheckboxModel(object, object["cls"]));
                    break;
                case dynamic_checkbox_group_model_1.DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP:
                    object["group"] = _this.fromJSON(object["group"]);
                    formModel.push(new dynamic_checkbox_group_model_1.DynamicCheckboxGroupModel(object, object["cls"]));
                    break;
                case dynamic_form_group_model_1.DYNAMIC_FORM_CONTROL_TYPE_GROUP:
                    object["group"] = _this.fromJSON(object["group"]);
                    formModel.push(new dynamic_form_group_model_1.DynamicFormGroupModel(object, object["cls"]));
                    break;
                case dynamic_input_model_1.DYNAMIC_FORM_CONTROL_TYPE_INPUT:
                    formModel.push(new dynamic_input_model_1.DynamicInputModel(object, object["cls"]));
                    break;
                case dynamic_radio_group_model_1.DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP:
                    formModel.push(new dynamic_radio_group_model_1.DynamicRadioGroupModel(object, object["cls"]));
                    break;
                case dynamic_select_model_1.DYNAMIC_FORM_CONTROL_TYPE_SELECT:
                    formModel.push(new dynamic_select_model_1.DynamicSelectModel(object, object["cls"]));
                    break;
                case dynamic_slider_model_1.DYNAMIC_FORM_CONTROL_TYPE_SLIDER:
                    formModel.push(new dynamic_slider_model_1.DynamicSliderModel(object, object["cls"]));
                    break;
                case dynamic_switch_model_1.DYNAMIC_FORM_CONTROL_TYPE_SWITCH:
                    formModel.push(new dynamic_switch_model_1.DynamicSwitchModel(object, object["cls"]));
                    break;
                case dynamic_textarea_model_1.DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA:
                    formModel.push(new dynamic_textarea_model_1.DynamicTextAreaModel(object, object["cls"]));
                    break;
                default:
                    throw new Error("unknown form control type with id \"" + object["id"] + "\" defined on JSON object");
            }
        });
        return formModel;
    };
    return DynamicFormService;
}());
DynamicFormService = __decorate([
    __param(0, core_1.Inject(forms_1.FormBuilder)),
    __param(1, core_1.Optional()), __param(1, core_1.Inject(forms_1.NG_VALIDATORS)),
    __param(2, core_1.Optional()), __param(2, core_1.Inject(forms_1.NG_ASYNC_VALIDATORS)),
    __metadata("design:paramtypes", [forms_1.FormBuilder,
        Array,
        Array])
], DynamicFormService);
exports.DynamicFormService = DynamicFormService;
